### PR TITLE
netcdf-cxx4: use mpi compile wrapper if needed.

### PR DESCRIPTION
**Pull request by tkameyama**
If netcdf-c was compiled with MPI, netcdf-cxx4 is compiled with mpicc and mpicxx.

### CODE REVIEW

**Comment by adamjstewart** [comment]
@WardF @skosukhin can you review this PR?

**Comment by tkameyama** [comment]
ping @adamjstewart @wardf @skosukhin

**Comment by whardier** [comment]
Why me?

**Comment by tkameyama** [comment]
> 
> 
> Why me?

Sorry, I misse user name.

File: var/spack/repos/builtin/packages/netcdf-cxx4/package.py [hunk 1]
```diff
@@ -56,6 +56,12 @@ def libs(self):
     def configure_args(self):
         config_args = []
 
+        if self.spec.satisfies('^mpi'):
```
Comment (skosukhin) [inline var/spack/repos/builtin/packages/netcdf-cxx4/package.py:60]: I would specify the condition like this (like in #16719):
```python
netcdf_c_spec = self.spec['netcdf-c']
if '+mpi' in netcdf_c_spec or '+parallel-netcdf' in netcdf_c_spec:
```
However, the current solution is also fine in my opinion.

Comment (tkameyama) [reply var/spack/repos/builtin/packages/netcdf-cxx4/package.py:60]: When netcdf-c is updated (For example new variant is added and the variant is needed MPI),
#16719 solution is needed to change netcdf-cxx.
But thi PR is not needed to change.

Comment (skosukhin) [reply var/spack/repos/builtin/packages/netcdf-cxx4/package.py:60]: The fact that `mpi` is somewhere in the dependency tree does not necesserily mean that we need to use MPI wrappers. For example, in theory (not in Spack yet), it is possible to build `netcdf-c~mpi ^hdf5+mpi` and I am not sure that we need to compile `netcdf-cxx` with mpi wrappers in this case.

Comment (tkameyama) [reply var/spack/repos/builtin/packages/netcdf-cxx4/package.py:60]: I see.
I Changed to look at variant of netcdf-c.

**Comment by skosukhin** [comment]
@tkameyama thank you. A few more comments:

1. I am not sure that we need to build with MPI wrappers at all. `netcdf-fortran`, for example, uses `mpi` at least in the test suite and in the examples. Therefore, it makes sense to build it with MPI wrappers. `netcdf-cxx4`, on the other hand, does not seem to use `mpi` directly, the dependencies on MPI libraries are transitive and are supposed to be resolved implicitly via rpaths. The case of static linking does not seem to be properly covered anyway (at least I got linking errors when tried to build `netcdf-cxx4 ^netcdf-c~mpi~shared`). Could you, please, provide an example when we need to use MPI wrappers?

2. There is a problem with `NEEDED` entries in `libnetcdf_c++4.so`: https://github.com/Unidata/netcdf-cxx4/issues/86. Basically, we can't simply link to the library: `g++ main.c -L/some/path -lnetcdf_c++4`. Instead, we always have to append `-lnetcdf` to the command. We can solve this in Spack. We can either:
    ```python
    config_args.append('LIBS=-lnetcdf')
    ```
    or wait until #17254 is merged and extend `flag_handler` with:
    ```python
    elif name == 'ldlibs':
        flags.append('-lnetcdf')
    ```
   Could you, please, take care of this? In any case, it makes sense to add the link to the issue as a comment: the issue might get resolved upstream in the future and the workaround might become redundant.

3. Currently, the package unconditionally depends on `automake`, `autoconf`, `libtool`, and `m4`. This is because we have `force_autoreconf = True`, which was introduced ages ago. Could you, please, check whether we need this for the newer version (I would simply remove it).

Of course, it's fine if you say no to any of this. I just thought that since you seem to be interested in the package, you might want to fix some other issues as well.

**Comment by tkameyama** [comment]
>  1. I am not sure that we need to build with MPI wrappers at all. `netcdf-fortran`, for example, 
> uses `mpi` at least in the test suite and in the examples. Therefore, it makes sense to build it with
>  MPI wrappers. `netcdf-cxx4`, on the other hand, does not seem to use `mpi` directly, the dependencies 
> on MPI libraries are transitive and are supposed to be resolved implicitly via rpaths. The case of static 
> linking does not seem to be properly covered anyway (at least I got linking errors when tried to build `netcdf-cxx4 ^netcdf-c~mpi~shared`). Could you, please, provide an example when we need to use MPI wrappers?

I faild to compile plugins/H5Zbzip2.c.
H5Zbzip2.c include hdf5.h, and hdf5.h include mpi.h when hdf5+mpi.
(And I use Fujitsu MPI, and this package's header information is incorrect. #17253)
I restore mpi wrapper condition to dependency baseed.

> 2. There is a problem with `NEEDED` entries in `libnetcdf_c++4.so`: 
> [Unidata/netcdf-cxx4#86](https://github.com/Unidata/netcdf-cxx4/issues/86). 
> Basically, we can't simply link to the library: `g++ main.c -L/some/path -lnetcdf_c++4`. 
> Instead, we always have to append `-lnetcdf` to the command. We can solve this in Spack. 

I think it's better to solve this problem by adding -lnetcdf to LIBS in the configure script.
I add patch for this.

>     3. Currently, the package unconditionally depends on `automake`, `autoconf`, `libtool`, and `m4`. This is because we have `force_autoreconf = True`, which was introduced ages ago. Could you, please, check whether we need this for the newer version (I would simply remove it).

Maybe we still need it.

**Comment by tkameyama** [comment]
@skosukhin 
> 2. There is a problem with `NEEDED` entries in `libnetcdf_c++4.so`: [Unidata/netcdf-cxx4#86](https://github.com/Unidata/netcdf-cxx4/issues/86). Basically, we can't simply link to the library: `g++ main.c -L/some/path -lnetcdf_c++4`. Instead, we always have to append `-lnetcdf` to the command. We can solve this in Spack. 

I understand this problem.

Nomally, If we want too use libnetcdf, we use AC_SEARCH_LIBS in configure.ac.
`AC_SEARCH_LIBS([nc_create], [netcdf], [], [])`
AC_SEARCH_LIBS search libnetcdf and add -lnetcdf to LIBS variable.

But currently when nc-config is found, configure.ac skip AC_SEARCH_LIBS, and LIBS variable is not updated.

libnetcdf_so.4.3.1.patch append output of `nc-config --libs` in LIBS variable.

**Comment by alalazo** [comment]
@skosukhin @tkameyama Is this something we still want or can we close the PR?

**Comment by alalazo** [comment]
Closing this PR as stale, but feel free to reopen if you want to continue working on it.

**Pull request closed without merging**